`Self` does not work in a field with serialize_with

Closely related to #1506, #1565, #1830.

```rust
use serde::{Serialize, Serializer};

#[derive(Serialize)]
struct Struct {
    #[serde(serialize_with = "vec_first_element")]
    vec: Vec<Self>,  // works if written as Vec<Struct>
}

fn vec_first_element<S>(vec: &Vec<Struct>, serializer: S) -> Result<S::Ok, S::Error>
where
    S: Serializer,
{
    vec.first().serialize(serializer)
}
```

```console
error[E0308]: mismatched types
 --> src/main.rs:3:10
  |
3 | #[derive(Serialize)]
  |          ^^^^^^^^^ expected struct `__SerializeWith`, found struct `Struct`
  |
  = note: expected reference `&Vec<__SerializeWith<'_>>`
             found reference `&Vec<Struct>`
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> src/main.rs:3:10
  |
3 | #[derive(Serialize)]
  |          ^^^^^^^^^ expected struct `Struct`, found struct `__SerializeWith`
  |
  = note: expected reference `&Vec<Struct>`
             found reference `&'__a Vec<__SerializeWith<'__a>>`
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/serde_derive/src/lib.rs
+++ b/serde_derive/src/lib.rs
@@ -79,8 +79,8 @@ mod try;
 
 #[proc_macro_derive(Serialize, attributes(serde))]
 pub fn derive_serialize(input: TokenStream) -> TokenStream {
-    let input = parse_macro_input!(input as DeriveInput);
-    ser::expand_derive_serialize(&input)
+    let mut input = parse_macro_input!(input as DeriveInput);
+    ser::expand_derive_serialize(&mut input)
         .unwrap_or_else(to_compile_errors)
         .into()
 }

--- a/serde_derive/src/ser.rs
+++ b/serde_derive/src/ser.rs
@@ -6,10 +6,14 @@ use bound;
 use dummy;
 use fragment::{Fragment, Match, Stmts};
 use internals::ast::{Container, Data, Field, Style, Variant};
-use internals::{attr, Ctxt, Derive};
+use internals::{attr, replace_receiver, Ctxt, Derive};
 use pretend;
 
-pub fn expand_derive_serialize(input: &syn::DeriveInput) -> Result<TokenStream, Vec<syn::Error>> {
+pub fn expand_derive_serialize(
+    input: &mut syn::DeriveInput,
+) -> Result<TokenStream, Vec<syn::Error>> {
+    replace_receiver(input);
+
     let ctxt = Ctxt::new();
     let cont = match Container::from_ast(&ctxt, input, Derive::Serialize) {
         Some(cont) => cont,

--- a/test_suite/tests/test_gen.rs
+++ b/test_suite/tests/test_gen.rs
@@ -7,6 +7,7 @@
 #![allow(
     unknown_lints,
     mixed_script_confusables,
+    clippy::ptr_arg,
     clippy::trivially_copy_pass_by_ref
 )]
 
@@ -735,6 +736,12 @@ fn test_gen() {
         #[serde(borrow = "'a")]
         f: mac!(Cow<'a, str>),
     }
+
+    #[derive(Serialize)]
+    struct Struct {
+        #[serde(serialize_with = "vec_first_element")]
+        vec: Vec<Self>,
+    }
 }
 
 //////////////////////////////////////////////////////////////////////////
@@ -807,4 +814,12 @@ where
 
 pub fn is_zero(n: &u8) -> bool {
     *n == 0
+}
+
+fn vec_first_element<T, S>(vec: &Vec<T>, serializer: S) -> StdResult<S::Ok, S::Error>
+where
+    T: Serialize,
+    S: Serializer,
+{
+    vec.first().serialize(serializer)
 }

